libass/ass_render.c

```diff
@@ -113,7 +111,13 @@ ASS_Renderer *ass_renderer_init(ASS_Library *library)
     priv->settings.shaper = ASS_SHAPING_SIMPLE;
 #endif
 
-  ass_init_exit:
+    if (!priv->cache.font_cache || !priv->cache.bitmap_cache || !priv->cache.composite_cache || !priv->cache.outline_cache ||
+        !priv->text_info.combined_bitmaps || priv->text_info.glyphs || !priv->text_info.lines || !priv->shaper) {
+fail:
```

Comment:
Jump into a conditioned block looks ugly, especially as next condition is basically a continuation of this block. I think it's better to move `fail:` at the function end, together with deallocation, `priv = NULL` and message. And in this condition to do `goto fail` similar to failures above. Maybe even better to split that monstrous condition into parts and move them near corresponding allocations.

Reply:
Done.